### PR TITLE
Generic tunnel

### PR DESCRIPTION
**Pull request by jdkanani**

### CODE REVIEW

File: contracts/tunnel/MessageTunnel.sol [hunk 1]
```diff
@@ -0,0 +1,201 @@
+pragma solidity ^0.6.6;
+
+
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {IStateSender} from "./root/StateSender/IStateSender.sol";
+import {RLPReader} from "./lib/RLPReader.sol";
+import {MerklePatriciaProof} from "./lib/MerklePatriciaProof.sol";
+import {ICheckpointManager} from "./root/RootChainManager/ICheckpointManager.sol";
```
Comment (arthcp) [inline contracts/tunnel/MessageTunnel.sol:8]: Need to fix the paths

File: contracts/tunnel/MockChildTunnel.sol [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    bytes32 public constant MSG1 = keccak256("MSG1");
+    bytes32 public constant MSG2 = keccak256("MSG2");
+
+    // MessageTunnel on L1 will get data from this event
+    event SendMessage(bytes);
+
+    // This method will be called by Matic chain internally, without transaction. 
+    // It's called `system-call`.
+    // It won't generate any event log since it's not a transaction based state change.
+    function receiveMessage(uint256, bytes calldata data) public only(STATE_SYNCER_ROLE) {
```
Comment (arthcp) [inline contracts/tunnel/MockChildTunnel.sol:21]: Change function name to `onStateReceive`

File: contracts/tunnel/MockChildTunnel.sol [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        (bytes32 syncType, bytes memory syncData) = abi.decode(
+            data,
+            (bytes32, bytes)
+        );
+
+        if (syncType == MSG1) {
+            // process syncData according to MSG1
+        } else if (syncType == MSG2) {
+            // process syncData according to MSG2
+        }
```
Comment (arthcp) [inline contracts/tunnel/MockChildTunnel.sol:31]: `syncType` is only present on child side, it can be confusing so lets keep it consistent

**Review by arthcp — CHANGES_REQUESTED** [review]

File: contracts/tunnel/BaseChildTunnel.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    // It's called `system-call`.
+    // It won't generate any event log since it's not a transaction based state change.
+    function onStateReceive(uint256, bytes memory data) public only(STATE_SYNCER_ROLE) {
+        processMessage(data);
+    }
+
+    // Send message from L2 to L1
+    function sendMessage(bytes memory data) public only(CHILD_SENDER_ROLE) {
+        emit SendMessage(data);
+    }
```
[the remark this answers is not in the document]
Comment (arthcp) [reply contracts/tunnel/BaseChildTunnel.sol:30]: Also lets remove the only role for this function. Basically, this function is called internally and all restriction etc. need not be implemented by base contract. All that can be done by the implementation contract.

File: contracts/tunnel/BaseRootTunnel.sol [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        only(DEFAULT_ADMIN_ROLE)
+    {
+        require(newChildTunnel != address(0x0), "MessageTunnel: INVALID_CHILD_TUNNEL_ADDRESS");
+        childTunnel = newChildTunnel;
+    }
+  
+    // send message from L1 to L2
+    function sendMessage(bytes calldata data) external only(SENDER_MANAGER_ROLE) {
+        stateSender.syncState(childTunnel, data);
+    }
```
Comment (arthcp) [inline contracts/tunnel/BaseRootTunnel.sol:81]: Same comment as above, lets make this internal and have implementation contract determine when it needs to be called

File: contracts/tunnel/BaseChildTunnel.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdkanani) [reply contracts/tunnel/BaseChildTunnel.sol:30]: Yeah then we need to make sure we convey this to docs.

**Review by arthcp — APPROVED** [review]

**Pull request merged**